File: #1220.count-vowels-permutation.cpp
class Solution {
public:
    int countVowelPermutation(int n) {

		long a = 1;
        long e = 1;
        long i = 1;
        long o = 1;
        long u = 1;
		int mod = 1e9 + 7;
		for (int j = 2; j <= n; j++) {
			long a2 = (e + u + i) % mod;
			long e2 = (a + i) % mod;
			long i2 = (e + o) % mod;
			long o2 = i % mod;
			long u2 = (o + i) % mod;
			a = a2;
			e = e2;
			i = i2;
			o = o2;
			u = u2;
		}
		return (a + e + i + o + u) % mod;
    }
};

File: #1220.count-vowels-permutation.js
/**
 * @param {number} n
 * @return {number}
 */
var countVowelPermutation = function(n) {
    const sum = (x,y) => (x + y) % (10 ** 9 + 7);
    const pre = [[1, 2, 4], [0, 2], [1, 3], [2], [2, 3]]; 
    const dp = [[0, 1], [0, 1], [0,1 ], [0,1 ], [0, 1]];                                
    for (let i = 2; i <= n; i++) {
        let k = i % 2;
        for (let j = 0; j < 5; j++) {
            dp[j][k] = pre[j].reduce((left, right) => 
                sum(left, dp[right][1-k]), 0);
        }
    }
    
    return dp.reduce((left, right) =>
        sum(left, right[n % 2]), 0);
};
